doc: Add bash dependency of lint tests

fjahr opened the pull request:
This issue has already been the topic of #24610.

macOS, sadly, does not ship with a bash version that is compatible with all lint tests since they now include the usage of `mapfile`. `mapfile` was introduced in `bash` 4.0 but macOS still uses 3.x because the newer version changed their license from GPL2 to GPL3. macOS has changed the default shell to `zsh` but that doesn't help since bash 3.x is still installed and will be used due to the shebang. Also, `zsh` does not have `mapfile` the same way `bash` has. Confusingly, the `zsh` version of the command does something completely different than the `bash` version.

Including guard code was rejected in #24610 and I guess this is probably reasonable since only a hand full of developers will ever run into this. But I think it should still be mentioned in the docs.

fjahr committed:
doc: Add bash dependency of lint tests

fjahr committed:
lint: Skip and warn in case of old bash version

co-authored-by: brunoerg <brunoely.gc@gmail.com>

jarolrod commented on test/README.md:
```diff
@@ -303,6 +303,15 @@ Use the `-v` option for verbose output.
 
 #### Dependencies
 
+##### Shell
```
You could changes this from Shell to Bash. This would help hint that a macOS user should switch from zhs to bash.

fjahr replied:
done, thanks!

jarolrod commented on test/README.md:
```diff
@@ -303,6 +303,15 @@ Use the `-v` option for verbose output.
+
+Several lint tests require a `bash` shell version 4 or higher to be available.
+
+Especially modern macOS versions (at least 10.15) still require installation of
```
I think the additions here could be condensed into:

```diff

diff --git a/test/README.md b/test/README.md
index 87a3acb76..cb96697e4 100644
--- a/test/README.md
+++ b/test/README.md
@@ -303,6 +303,13 @@ Use the `-v` option for verbose output.
 
 #### Dependencies
 
+##### Bash
+
+Several lint tests require a `bash` shell version 4 or higher.
+macOS systems require upgrading the system provided `bash`.
+
+##### Python
+
```

fjahr replied:
done, thanks!

jarolrod reviewed (commented):
concept ack

RandyMcMillan commented:
concept ACK

RandyMcMillan commented:
In a related issue - the default 'sed' version fails on MacOS...

https://github.com/bitcoin/bitcoin/pull/24159

It may be useful to add 'gsed' to the list of homebrew dependancies.

brunoerg commented:
Since #24610 has been closed, concept ACK.

MarcoFalke commented:
If there are systems that ship an outdated bash, it may be better to reopen #24610 instead. A simple automated check is faster and more friendly than documentation, which needs to be searched and interpreted by a human.

brunoerg commented:
> If there are systems that ship an outdated bash, it may be better to reopen #24610 instead. A simple automated check is faster and more friendly than documentation, which needs to be searched and interpreted by a human.

I agree. It would be more friendly.

laanwj commented:
I'd prefer one external scripting dependency (Python), but if we insist on modern bash-isms, could do both #24610 and this. Both documenting and detecting dependencies makes sense.

fjahr commented:
I have added a commit that adds the `bash` warning to all the linters that use `mapfile`. I had written it yesterday but ended up not pushing it because of the negative sentiment in #24610. If there is interest in adding this together with the docs we can do it here. I added @brunoerg as co-author of course.

I agree switching to `python` is the better solution and happy to help with that as well. Then the only question is if it still makes sense to still add this now.

fjahr commented:
Also addressed @jarolrod 's comments on the doc changes, thanks for reviewing!

brunoerg reviewed (approved):
crACK d8cc7993aa2985e785acc09c7a1b64989db5877a

MarcoFalke commented:
NACK. Agree with laanwj

fjahr commented:
Thanks for the reviews but I will close this for now since I have already converted almost all of the offending code to Python and I will finish the rest today. I have not seen too many compatibility issues myself previously but I think converting to Python makes this also easier to maintain because more developers are available.

See #24766 and #24778.

laanwj commented:
> Thanks for the reviews but I will close this for now since I have already converted almost all of the offending code to Python and I will finish the rest today.

That's awesome! Thank you.
